relay-server/src/actors/spooler/mod.rs

```diff
@@ -501,18 +394,330 @@ impl BufferService {
         }
     }
 
-    /// Checks if the spool still has space and sets `is_disk_full` flag accordingly.
-    async fn refresh_spool_state(&mut self) {
-        let Some(BufferSpoolConfig {
+    /// Dequeues the envelopes from the on-disk spool and send them to the provided `sender`.
+    ///
+    /// The keys for which the envelopes could not be fetched, send back to `ProjectCache` to merge
+    /// back into index.
+    async fn dequeue(
+        &self,
+        project_key: ProjectKey,
+        keys: &mut Vec<QueueKey>,
+        sender: mpsc::UnboundedSender<ManagedEnvelope>,
+        services: &Services,
+    ) {
+        let mut unused_keys = BTreeSet::new();
+        while let Some(key) = keys.pop() {
+            // If the error with a key is returned we must save it for the next iterration.
+            if let Err(key) = self.delete_and_fetch(key, &sender, services).await {
+                unused_keys.insert(key);
+            };
+        }
+        services
+            .project_cache
+            .send(UpdateBufferIndex::new(project_key, unused_keys))
+    }
+
+    /// Estimates the db size by multiplying `page_count * page_size`.
+    async fn estimate_spool_size(&self) -> Result<i64, BufferError> {
+        let size: i64 = sql::current_size()
+            .fetch_one(&self.db)
+            .await
+            .and_then(|r| r.try_get(0))
+            .map_err(BufferError::FileSizeReadFailed)?;
+
+        relay_statsd::metric!(histogram(RelayHistograms::BufferDiskSize) = size as u64);
+        Ok(size)
+    }
+
+    /// Returns `true` if the maximum size is reached, `false` otherwise.
+    async fn is_full(&self) -> Result<bool, BufferError> {
+        let current_size = self.estimate_spool_size().await?;
+        Ok(current_size as usize >= self.max_disk_size)
+    }
+
+    /// Returns `true` if the spool is empty, `false` otherwise.
+    async fn is_empty(&self) -> Result<bool, BufferError> {
+        let is_empty = sql::select_one()
+            .fetch_optional(&self.db)
+            .await
+            .map_err(BufferError::FetchFailed)?
+            .is_none();
+
+        Ok(is_empty)
+    }
+
+    /// Enqueues data into on-disk spool.
+    async fn enqueue(
+        &self,
+        key: QueueKey,
+        managed_envelope: ManagedEnvelope,
+    ) -> Result<(), BufferError> {
+        let received_at = managed_envelope.received_at().timestamp_millis();
+        sql::insert(
+            key,
+            managed_envelope.into_envelope().to_vec().unwrap(),
+            received_at,
+        )
+        .execute(&self.db)
+        .await
+        .map_err(BufferError::InsertFailed)?;
+
+        relay_statsd::metric!(counter(RelayCounters::BufferWrites) += 1);
+        Ok(())
+    }
+}
+
+/// The state which defines the [`BufferService`] behaviour.
+#[derive(Debug)]
+enum BufferState {
```

Comment:
Even with the PR description, I find these names confusing. What do you think about the following suggestions?

- `MemoryOnly`.
- `MemoryFirstSpoolStandby`.
- `MemoryFirstSpoolRead`.
- `SpoolOnly`.

Reply:
I agree with the sentiment, but I find the suggestions just as confusing. Maybe naming is not the problem? E.g., maybe `Memory` and `Standby` don't have to be separate states?

Reply:
I find the suggestions even more confusing and would prefer to have shorter names. 
`Memory` here is only when the disk is not configured, which makes it easier to work with. If the `Standby` state is re-used, I will have to wrap the disk in `Option`, which also will have to be treated differently?